### PR TITLE
feat: add maxValuesPerDimension cap to streamingPartitionsSpec

### PR DESCRIPTION
**Pull request by jaykanakiya**
### Description
This change adds  maxValuesPerDimension, an optional safety cap on distinct values recorded per dimension per segment in the dim_value_set shard spec added in https://github.com/apache/druid/pull/19571 for an optional guard rail.

#### Release note

<hr>

##### Key changed/added classes in this PR
 * `SeekableStreamIndexTaskRunner.java`
 * `StreamingPartitionsSpec.java`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Review by FrankChen021 — COMMENTED** [review]
I have reviewed the code for correctness, edge cases, concurrency, and integration risks; no issues found.

Reviewed 4 of 4 changed files.

---

This is an automated review by Codex GPT-5.5

File: docs/ingestion/kafka-ingestion.md [hunk 1]
```diff
@@ -273,6 +273,7 @@ This enables segment pruning for streaming-ingested data without waiting for com
 
 - Only string-typed dimensions are currently supported.
 - Use only low-to-medium cardinality dimensions (for example, `tenant_id`, `region`, `environment`). High-cardinality dimensions bloat segment metadata with no pruning benefit.
+- Set `maxValuesPerDimension` as a safety cap if a tracked dimension may unexpectedly grow high-cardinality. When a segment's observed distinct values for a dimension exceed the cap, that dimension is omitted from the segment's stamped filter map: pruning is disabled for that dimension on that segment, but other tracked dimensions continue to prune as normal. Default is unset (uncapped).
```
Comment (abhishekrb19) [inline docs/ingestion/kafka-ingestion.md]:
```suggestion
- Set `maxValuesPerDimension` as a safety cap if a tracked dimension may unexpectedly grow high-cardinality. When a segment's observed distinct values for a dimension exceed the cap, that dimension is omitted from the segment's stamped filter map: pruning is disabled for that dimension on that segment, but other tracked dimensions continue to prune as normal. Default is unlimited (uncapped).
```

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/StreamingPartitionsSpec.java [hunk 2]
```diff
@@ -41,13 +42,29 @@
...
   )
   {
     this.partitionDimensions = partitionDimensions == null ? Collections.emptyList() : partitionDimensions;
+    if (maxValuesPerDimension != null) {
+      Preconditions.checkArgument(
+          maxValuesPerDimension > 0,
+          "maxValuesPerDimension must be > 0, got [%s]",
+          maxValuesPerDimension
+      );
+    }
```
Comment (abhishekrb19) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/StreamingPartitionsSpec.java]: Nit: Could you switch this to `DruidException`?

File: indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunnerTest.java [hunk 3]
```diff
@@ -580,6 +580,103 @@ public void testFeatureOffReturnsSegmentUnchanged() throws Exception
...
+        Arrays.asList("tenant_a", "tenant_b", "tenant_c"),
+        ((DimensionValueSetShardSpec) annotated.getShardSpec()).getPartitionDimensionValues().get("tenant")
+    );
+  }
+
+  /** Over-cap: dim is omitted from the filter map; segment still gets a DimensionValueSetShardSpec. */
+  @Test
+  public void testCapExceededOmitsDimensionFromFilterMap() throws Exception
+  {
+    final TestSeekableStreamIndexTaskRunner runner = createRunner(
```
Comment (abhishekrb19) [inline indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunnerTest.java:610]: Could we also add a couple of embedded tests here to verify the behavior end-to-end: https://github.com/apache/druid/blob/master/extensions-core/kafka-indexing-service/src/test/java/org/apache/druid/indexing/kafka/simulate/EmbeddedDimensionValueSetShardSpecTest.java#L112

**Review by abhishekrb19 — COMMENTED** [review]
Thanks @jaykanakiya! Some comments below.

File: docs/ingestion/kafka-ingestion.md (hunk 1 above)
Comment (jaykanakiya) [reply docs/ingestion/kafka-ingestion.md]: Updated the text.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/StreamingPartitionsSpec.java (hunk 2 above)
Comment (jaykanakiya) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/StreamingPartitionsSpec.java]: Updated

File: indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunnerTest.java (hunk 3 above)
Comment (jaykanakiya) [reply indexing-service/src/test/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunnerTest.java:610]: Added embedded test

**Review by abhishekrb19 — APPROVED** [review]
